plugins/procedure/init.inc.php

```diff
@@ -0,0 +1,246 @@
+<?php
+/*
+ * Copyright (C) 2019, Daniel Haslinger <creo+oss@mesanova.com>
+ * This program is free software licensed under the terms of the GNU General Public License v3 (GPLv3).
+ */
+
+class PLUGIN_PROCEDURE
+{
+    private $object_broker;
+    private $classname;
+    const ACL_MODE = 'none';    // white, black, none
+
+    public function __construct($object_broker)
+    {
+        $this->classname = strtolower(static::class);
+
+        $this->object_broker = $object_broker;
+        $object_broker->plugins[] = $this->classname;
+        error_log($this->classname . ": starting up");
+
+        //$this->object_broker->instance['api_routing']->register("bootup", $this->classname, "Procedure to start the space");
+        $this->object_broker->instance['api_routing']->register("teardown", $this->classname, "Procedure to shutdown the space");
+        //$this->object_broker->instance['api_routing']->helptext("bootup", "", "Procedure to start the space");
+        $this->object_broker->instance['api_routing']->helptext("teardown", "", "Procedure to shutdown the space");
+    }
+
+
+    public function __destruct()
+    {
+
+    }
+
+
+    public function get_acl_mode()
+    {
+        return self::ACL_MODE;
+    }
+
+    private function send_to_user($text, $inlineKB=null)
+    {
+        $senderid = $GLOBALS['layer7_stanza']['message']['from']['id'];
+        return $this->object_broker->instance['api_telegram']->send_message($senderid, $text, $inlineKB);
+    }
+
+    private function delete_msg($msgId)
+    {
+        $senderid = $GLOBALS['layer7_stanza']['message']['from']['id'];
+        $this->object_broker->instance['api_telegram']->delete_message($senderid, $msgId);
+    }
+
+    public function process($trigger)
+    {
+        error_log($this->classname . ": processing trigger $trigger");
+
+        $chatid = $GLOBALS['layer7_stanza']['message']['chat']['id'];
+        $senderid = $GLOBALS['layer7_stanza']['message']['from']['id'];
+
+        $payload = str_replace('/' . $trigger, '', $GLOBALS['layer7_stanza']['message']['text']);
+        $payload = trim($payload);
+        error_log($this->classname . ": payload \"$payload\"");
+
+        $herald_ok = $this->object_broker->instance['api_routing']->acl_check_list($senderid, "plugin_heralding", "white");
+
+        switch($trigger)
+        {
+            case "bootup":
+                break;
+
+            case "teardown":
+                if( empty($payload) )
+                {
+                    $payload = "start";
+                }
+                $payload_arr = preg_split("/\s+/", $payload);
+                if( count($payload_arr) <= 0 )
+                {
+                    return false;
+                }
+                switch($payload_arr[0])
```

Comment:
I solved this problem exactly the same way you're doing it here when I wrote the first spacebot,
so let's leave it like this for now, but for the future:
we should probably not hardcode the whole procedure into the source code but pull the individual tasks and button definitions from e.g. a JSON file